Add depth estimation pipeline

nandwalritik opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.
* I tried debugging
Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #18446 


## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?
@NielsRogge @Narsil 
<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

## Error While using the pipeline
```
pipe = pipeline("depth-estimation")
No model was supplied, defaulted to Intel/dpt-large and revision e93beec (https://huggingface.co/Intel/dpt-large).
Using a pipeline without specifying a model name and revision in production is not recommended.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nandwalritik/nandwalritik/transformers/src/transformers/pipelines/__init__.py", line 670, in pipeline
    framework, model = infer_framework_load_model(
  File "/home/nandwalritik/nandwalritik/transformers/src/transformers/pipelines/base.py", line 257, in infer_framework_load_model
    model = model_class.from_pretrained(model, **kwargs)
  File "/home/nandwalritik/nandwalritik/transformers/src/transformers/models/auto/auto_factory.py", line 445, in from_pretrained
    model_class = _get_model_class(config, cls._model_mapping)
  File "/home/nandwalritik/nandwalritik/transformers/src/transformers/models/auto/auto_factory.py", line 359, in _get_model_class
    supported_models = model_mapping[type(config)]
  File "/home/nandwalritik/nandwalritik/transformers/src/transformers/models/auto/auto_factory.py", line 565, in __getitem__
    return self._load_attr_from_module(model_type, model_name)
  File "/home/nandwalritik/nandwalritik/transformers/src/transformers/models/auto/auto_factory.py", line 579, in _load_attr_from_module
    return getattribute_from_module(self._modules[module_name], attr)
  File "/home/nandwalritik/nandwalritik/transformers/src/transformers/models/auto/auto_factory.py", line 539, in getattribute_from_module
    return getattribute_from_module(transformers_module, attr)
  File "/home/nandwalritik/nandwalritik/transformers/src/transformers/models/auto/auto_factory.py", line 539, in getattribute_from_module
    return getattribute_from_module(transformers_module, attr)
  File "/home/nandwalritik/nandwalritik/transformers/src/transformers/models/auto/auto_factory.py", line 539, in getattribute_from_module
    return getattribute_from_module(transformers_module, attr)
  [Previous line repeated 982 more times]
  File "/home/nandwalritik/nandwalritik/transformers/src/transformers/models/auto/auto_factory.py", line 538, in getattribute_from_module
    transformers_module = importlib.import_module("transformers")
  File "/home/nandwalritik/anaconda3/envs/hftfSwinDev/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1004, in _find_and_load
  File "<frozen importlib._bootstrap>", line 157, in __enter__
  File "<frozen importlib._bootstrap>", line 183, in _get_module_lock
  File "<frozen importlib._bootstrap>", line 59, in __init__
RecursionError: maximum recursion depth exceeded while calling a Python object
```

nandwalritik committed:
Add initial files for depth estimation pipelines

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18618). All of your documentation changes will be reflected on that endpoint.

nandwalritik commented:
Hi @NielsRogge I tried debugging to resolve above issue, I found that in `src/transformers/models/auto/auto_factory.py` in line 462  `model_class = _get_model_class(config, cls._model_mapping)` when I logged `cls._model_mapping` I recieved
```
OrderedDict([(<class 'transformers...PTConfig'>, <?>), (<class 'transformers...PNConfig'>, <?>)])
<error>:
Traceback (most recent call last):
```
Can you guide me to resolve this error.

Narsil commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -448,6 +448,13 @@
+MODEL_FOR_DEPTH_ESTIMATION_MAPPING_NAMES = OrderedDict(
+    [
+        # Model for depth estimation mapping
+        ("dpt", "DptForDepthEstimation"),
```
```suggestion
        ("dpt", "DPTForDepthEstimation"),
```

This is your bug, `Dpt` does not exist, so it loops trying to find this name.

Narsil commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -448,6 +448,13 @@
+MODEL_FOR_DEPTH_ESTIMATION_MAPPING_NAMES = OrderedDict(
+    [
+        # Model for depth estimation mapping
+        ("dpt", "DptForDepthEstimation"),
+        ("glpn", "GlpnForDepthEstimation"),
```
```suggestion
        ("dpt", "DPTForDepthEstimation"),
        ("glpn", "GLPNForDepthEstimation"),
```

nandwalritik replied:
Thanks It resolved the issue.

Narsil reviewed (commented):
Great PR, you also need to add

```python
+    def _sanitize_parameters(self, **kwargs):
+        return {}, {}, {}
+
```
To the pipeline, it's not used but the base class expects this method to exist. (Here there's not parameters declared so it's quite easy.

This weird method exist to allow parameters to be defined both at definition time or call tiime

```
pipe = pipeline(model=model, myargs=1)
data = pipe(image)
# or
pipe = pipeline(model=model)
data = pipe(image, myargs=1)

```

Cheers !
Otherwise LGTM.

Why do you output 3 different images ? That sounds like a lot.

The image I can understand, the predicted depth is defined in what unit ? Is it noisy hence the interpolation ?
IMO that seems like something to be left to the user to decide what to do.

In general I think a pure image would be nice (to be a bit more general) but I can understand that the loss of precision might be harmful, do you mind sharing how you use those numbers ? Maybe we could output an other time of image that doesn't loose information (keeping f32 pixel)

Wdyt ?

nandwalritik committed:
Add test file for depth estimation pipeline

nandwalritik committed:
Update model mapping names

nandwalritik committed:
Add updates for depth estimation output

nandwalritik commented:
> Great PR, you also need to add
> 
> ```python
> +    def _sanitize_parameters(self, **kwargs):
> +        return {}, {}, {}
> +
> ```
> 
> To the pipeline, it's not used but the base class expects this method to exist. (Here there's not parameters declared so it's quite easy.
> 
> This weird method exist to allow parameters to be defined both at definition time or call tiime
> 
> ```
> pipe = pipeline(model=model, myargs=1)
> data = pipe(image)
> # or
> pipe = pipeline(model=model)
> data = pipe(image, myargs=1)
> ```
> 
> Cheers ! Otherwise LGTM.
> 
> Why do you output 3 different images ? That sounds like a lot.
> 
> The image I can understand, the predicted depth is defined in what unit ? Is it noisy hence the interpolation ? IMO that seems like something to be left to the user to decide what to do.
> 
> In general I think a pure image would be nice (to be a bit more general) but I can understand that the loss of precision might be harmful, do you mind sharing how you use those numbers ? Maybe we could output an other time of image that doesn't loose information (keeping f32 pixel)
> 
> Wdyt ?
I just saw `DPT`'s depth estimation example and added these three outputs. I have removed the interpolation one and In the output I have kept only the `predicted_depth` (which is a `tensor`) and `depth` (which is the PIL `Image` object).
Let me know if I should remove the `predicted_depth` also.

Narsil reviewed (commented):
LGTM.

Aside from the missing generic test. 
Would you like us to write it  ? (It's not too hard to write I think).

My biggest question is why `depth-estimation` needs it's own pipeline.

I understand in the robotics context you really want something that's not an image, but really a matrix with depth (expressed iin meters I guess).
This reason only makes me ok with the idea of creating this pipeline as is.

But ideally it would be nice if the output was an image (so it could become `image-generation`) with either an output that would be trivial to transform into a depth map or a parameter.

@mishig25 maybe for advice ideas on the widget that could be different than an image/generation here ?

In any case, expect if anyone has better ideas counterarguments, we should just merge this right now (after having the generic test), and maybe later convert this to an alias if we decide to merge.

nandwalritik committed:
Add generic test

nandwalritik commented:
> ### Review required
> At least 1 approving review is required by reviewers with write access. [Learn more.](https://docs.github.com/articles/about-pull-request-reviews/)
> ** 1 pending reviewer **

By generic test did you mean `run_pipeline_test` ? Let me know if it's other than this, I have added `run_pipeline_test` for now.
Also In CI `test_pipelines_depth_estimation` is failing can you help me with that ?

Narsil committed:
Hopefully fixing the tests.

nandwalritik committed:
Check if test passes

nandwalritik committed:
Add make fixup and make fix-copies changes after rebase with main

NielsRogge commented:
Hi @nandwalritik, could you revive this PR by rebasing with the main branch?

nandwalritik committed:
Rebase with main

nandwalritik commented:
> Hi @nandwalritik, could you revive this PR by rebasing with the main branch?

Done.

Narsil commented:
@nandwalritik Do you want to help to get the PR green ?

nandwalritik commented:
> @nandwalritik Do you want to help to get the PR green ?

@Narsil Yeah please , I tried but I was not able to make the test cases pass.

Narsil committed:
Fixing up depth pipeline.

Narsil commented on docs/source/en/main_classes/pipelines.mdx:
```diff
@@ -25,11 +25,8 @@ There are two categories of pipeline abstractions to be aware about:
-  - [`DocumentQuestionAnsweringPipeline`]
-=======
   - [`DepthEstimationPipeline`]
->>>>>>> Add initial files for depth estimation pipelines
```
This was a leftover from a bad rebase I think, I removed those lines and put back the documentation in alphabetical order.

Narsil commented on docs/source/en/main_classes/pipelines.mdx:
```diff
@@ -348,20 +345,17 @@ That should enable you to do all the custom code you want.
     - all
-
-=======
-### DepthEstimationPipeline
```
Same here.

Narsil commented on src/transformers/pipelines/depth_estimation.py:
```diff
@@ -1,7 +1,6 @@
 from typing import List, Union
 
 import numpy as np
-import torch
```
`torch` is not a mandatory dependency, we need to import it only when it's OK to do so.

Narsil commented on tests/pipelines/test_pipelines_depth_estimation.py:
```diff
@@ -32,6 +30,9 @@
 
 
+if is_torch_available():
+    import torch
```
Same in tests.

Narsil reviewed (commented):
Added some comments on how I fixed the CI for you.

Narsil committed:
This is not used anymore.

Narsil commented on tests/pipelines/test_pipelines_depth_estimation.py:
```diff
@@ -17,15 +17,7 @@
-    require_vision,
-    slow,
-)
+from transformers.testing_utils import nested_simplify, require_tf, require_timm, require_torch, require_vision, slow
```
Since you create your PR and today we removed `is_pipeline_test` as we're now passing on these tests based on directory not by ENV variable.

Narsil committed:
Fixing the test. `Image` is a module `Image.Image` is the type.

Narsil commented on tests/pipelines/test_pipelines_depth_estimation.py:
```diff
@@ -56,7 +56,7 @@ def get_test_pipeline(self, model, tokenizer, feature_extractor):
     def run_pipeline_test(self, depth_estimator, examples):
         outputs = depth_estimator("./tests/fixtures/tests_samples/COCO/000000039769.png")
-        self.assertEqual({"predicted_depth": ANY(torch.Tensor), "depth": ANY(Image)}, outputs)
+        self.assertEqual({"predicted_depth": ANY(torch.Tensor), "depth": ANY(Image.Image)}, outputs)
```
`Image` is a module, `Image.Image` is the class we need to check against.

nandwalritik commented:
> Added some comments on how I fixed the CI for you.

Thanks I will look at them.

Narsil commented:
@sgugger for final review.

sgugger reviewed (approved):
LGTM! Thanks a lot for your work on this!

nandwalritik committed:
Update docs/source/en/main_classes/pipelines.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

sgugger commented:
Yes, those are flaky tests.
